### ISSUE TITLE
IPBan memory usage on windows server 2025

### ISSUE THREAD

**Issue by fhm20** [DigitalRuby/IPBan#316]
I'm testing IPBan on Windows Server 2025.

Server1: W2025 without IPBan. The process svchost.exe -k LocalServiceNoNetworkFirewall -p uses 6MB of RAM.
Server2: W2025 with IPBan 1.8.1 or 2.0.1. The process svchost.exe -k LocalServiceNoNetworkFirewall -p uses 2.1GB of RAM. Uptime: 15 days.
Server3: W2022 with IPBan 1.8.1. The process svchost.exe -k LocalServiceNoNetworkFirewall -p uses 22MB of RAM. Uptime: 30 days.
The RAM usage of svchost.exe -k LocalServiceNoNetworkFirewall on W2025 increases every day and does not release memory. In two days, it increased by 400MB.
I have the same 3 rules, "IPBan_GlobalBlacklist," each with 1000 IPs (3000 IPs in total), on all 3 servers.

This issue does not occur on W2022. Could you check if there is a memory problem with IPBan or if it is an issue with the firewall of Windows Server 2025?

Thanks

**Comment by jjxtra** [comment]
Out of curiosity, does IPBanProDatacenter have the same issue? There's a free trial you can install to test: https://ipban.com/products/ipban-pro-datacenter-free-trial-download/

This one doesn't use Windows firewall

**Comment by fhm20** [comment]
I installed version ipban 3.0 (.net9) and the final version of Windows Server 2025 (previously, I had Windows Server Preview), and the error persists. In two days, the svchost process increased to 500MB of RAM on two different machines. I’m going to try IPBanProDatacenter now.

**Comment by fhm20** [comment]
With the datacenter version, memory consumption is correct.

DigitalRuby.IPBanProDatacenter.exe  42.6 MB RAM
DigitalRuby.IPBanProWFPEvents.exe 2.4 MB RAM 
svchost.exe -k LocalServiceNoNetworkFirewall 10.3 MB RAM

IPBan 3.0
DigitalRuby.IPBan 34 MB RAM
svchost.exe -k LocalServiceNoNetworkFirewall 850 MB RAM

**Comment by jjxtra** [comment]
99% sure this is a problem with Windows firewall then. Will have to investigate what changed with COM or this implementation in server 2025. Could be a MSFT bug ...

**Comment by Maddymask** [comment]
> 99% sure this is a problem with Windows firewall then. Will have to investigate what changed with COM or their implementation in server 2025. Could be a MSFT bug ...

I'm just about to install your product on Windows Server 2025 on production. Is it not recommended to do this yet?

**Comment by jjxtra** [comment]
You can use it in production on a 2025 window server, however, you’ll need to set up a scheduled task to restart the service every 12 hours.

**Comment by jjxtra** [comment]
It’s an ugly hack, but my solution for this will be to do just exactly that, detect if it is running on window server 2025, and if so, the service will just restart itself when memory usage gets above a certain threshold.

**Comment by jjxtra** [comment]
I forgot to ask, does restarting ipban clear the memory usage of the firewall service @fhm20 ?

**Comment by fhm20** [comment]
No. If I restart the service, the memory is not released. I opened a case with Microsoft. I'm waiting for their response.

**Comment by jjxtra** [comment]
> > 99% sure this is a problem with Windows firewall then. Will have to investigate what changed with COM or their implementation in server 2025. Could be a MSFT bug ...
> 
> I'm just about to install your product on Windows Server 2025 on production. Is it not recommended to do this yet?

Unfortunately not recommended until memory leaks can be fixed in Windows

**Comment by jjxtra** [comment]
This Win32 api call has had a memory leak for years, too: `FwpmNetEventEnum2`

To work around it and surface firewall events properly in IPBan Pro, I had to write a separate executable, read the std out from the primary service, then kill the child process periodically as it filled up with RAM. You'd think calling `FwpmFreeMemory0` would actually free the memory, but alas.

All this is to say this wouldn't be the first time a memory leak crept it's way into Microsoft's firewall code.

**Comment by jjxtra** [comment]
@fhm20 Does restarting the LocalServiceNoNetworkFirewall service clear the memory usage?

**Comment by fhm20** [comment]
I can't restart the process cleany. I have to kill the process, which causes the server to lose connectivity for a few seconds. After that, the memory is released. Microsoft technical support is still investigating the issue.

**Comment by jjxtra** [comment]
What if the ipban service restarted this process daily at 3am for win2025 only, would that be acceptable? Or just wait for Microsoft response?

**Comment by fhm20** [comment]
It's not enough to restart the IPBan service to clean the memory. I would need to kill the svchost process, but I can't do that at night either. Would it be possible for you to adapt the DigitalRuby.IPBanProWFPEvents.exe to the free version?
If not, I’ll have to stop using IPBan on Windows 2025 until Microsoft fixes the issue.

**Comment by jjxtra** [comment]
Not restarting the IPBan service, LocalServiceNoNetworkFirewall. That should free up the memory while causing just a few seconds blip, if I understand your prior comments right.

**Comment by jjxtra** [comment]
Wow they actually found the problem. Nice. Closing, and thanks for reporting.
